loopy/schedule/__init__.py

```diff
@@ -645,6 +650,145 @@ def last_entered_loop(self):
         else:
             return None
 
+# }}}
+
+
+def get_insns_in_topologically_sorted_order(kernel):
+    from pytools.graph import compute_topological_order
+
+    rev_dep_map = {insn.id: set() for insn in kernel.instructions}
+    for insn in kernel.instructions:
+        for dep in insn.depends_on:
+            rev_dep_map[dep].add(insn.id)
+
+    def key(insn_id):
+        # negative of insn.priority because
+        # pytools.graph.compute_topological_order schedules the nodes with
+        # lower 'key' in case of a tie.
+        return (-kernel.id_to_insn[insn_id].priority, insn.id)
+
+    ids = compute_topological_order(rev_dep_map, key=key)
+    return [kernel.id_to_insn[insn_id] for insn_id in ids]
+
+
+# {{{ schedule_as_many_run_insns_as_possible
+
+def schedule_as_many_run_insns_as_possible(sched_state, template_insn):
+    """
+    Returns an instance of :class:`loopy.schedule.SchedulerState`, by appending
+    all reachable instructions that are similar to *template_insn*. We define
+    two instructions to be similar if:
+
+    * Both have the same type.
+    * Both are within the same set of non-parallel inames.
+    * Both belong to the same groups.
+    * Both conflict with the same groups.
+    """
+
+    # {{{ bail when implementation is unsupported
+
+    next_preschedule_item = (
+        sched_state.preschedule[0]
+        if sched_state.preschedule
+        else None)
+
+    if isinstance(next_preschedule_item, (CallKernel, ReturnFromKernel,
+            Barrier, EnterLoop, LeaveLoop)):
+        return sched_state
+
+    if not sched_state.within_subkernel:
+        # cannot schedule RunInstructions when not in subkernel
+        return sched_state
+
+    # }}}
+
+    preschedule = sched_state.preschedule[:]
+    have_inames = template_insn.within_inames - sched_state.parallel_inames
+    toposorted_insns = sched_state.insns_in_topologically_sorted_order
+
+    # {{{ helpers
+
+    def next_preschedule_insn_id():
+        return (next(iter(sched_item_to_insn_id(preschedule[0])), None)
+                if sched_state.preschedule
+                else None)
+
+    def is_similar_to_template(insn):
+        if type(insn) != type(template_insn):
+            return False
+        if ((insn.within_inames - sched_state.parallel_inames)
+                != have_inames):
+            return False
+        if insn.groups != template_insn.groups:
+            return False
+        if insn.conflicts_with_groups != template_insn.conflicts_with_groups:
+            return False
+
+        return True
+
+    # }}}
+
+    # select the top instructions in toposorted_insns only which have active
+    # inames corresponding to those of sched_state
+    newly_scheduled_insn_ids = []
+    ignored_unscheduled_insn_ids = set()
+
+    # new_toposorted_insns: unscheduled insns in a topologically sorted order
+    new_toposorted_insns = []
+
+    for insn in toposorted_insns:
```

Comment:
This keeps running through the entire topologically sorted list, although once we've had to leave some instructions out, I think the likelihood that we'll find one that we can schedule drops considerably. When it's forced to skip O(N) times (but is that likely?), the algorithm is likely to be quadratic, as it has to examine the tail of the list repeatedly.

Basically, I'm proposing that we might be better off bailing out after a few skips. (But you have more experience with this up close, so I'd be happy to hear your thoughts.)

Reply:
That was a slightly sub-optimal decision, as I couldn't find a cleaner solution.

In 3f5f651b I have introduced a heuristic to bail the process of scheduling many instructions at once as soon as we have ignored 5 instructions.

Reply:
Grammar nit: the correct usage in this case is "to bail out of sth." not "to bail sth."